flocker/ca/_ca.py

```diff
@@ -51,138 +66,238 @@ def __init__(self, keypair):
 
     :ivar FilePath path: A ``FilePath`` representing the absolute path of
-        a directory containing the CRT and KEY files.
-    :ivar Certificate certificate: A signed certificate, populated only by
-        loading from ``path``.
-    :ivar FlockerKeyPair keypair: A private/public keypair, populated only by
-        loading from path.
-    """
-    certificate = field(mandatory=True, initial=None)
-    keypair = field(mandatory=True, initial=None)
+        a directory containing the certificate and key files.
+    :ivar Certificate certificate: A signed certificate.
+    :ivar FlockerKeyPair keypair: A private/public keypair.
+    """
     path = field(mandatory=True)
+    certificate = field(mandatory=True)
+    keypair = field(mandatory=True)
 
-    @classmethod
-    def from_path(cls, path):
+    def write_credential_files(self, key_filename, certificate_filename):
         """
-        :param FilePath path: Directory where private key and certificate are
-            stored.
+        Write PEM encoded certificate and private key files for this credential
+        instance.
+
+        :param str key_filename: The name of the private key file to write,
+            e.g. "cluster.key"
+        :param str certificate_filename: The name of the certificate file to
+            write, e.g. "cluster.crt"
         """
-        if not path.isdir():
+        key_path = self.path.child(key_filename)
+        cert_path = self.path.child(certificate_filename)
+        original_umask = os.umask(0)
+        mode = 0o600
+        try:
+            with os.fdopen(os.open(
+                cert_path.path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, mode
+            ), b'w') as cert_file:
+                cert_file.write(self.certificate.dumpPEM())
+            try:
+                with os.fdopen(os.open(
+                    key_path.path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, mode
+                ), b'w') as key_file:
+                    key_file.write(
+                        self.keypair.keypair.dump(crypto.FILETYPE_PEM))
+            except (IOError, OSError) as e:
+                code, failure = e
+                raise PathError(
+                    b"Unable to write private key file.",
+                    e.filename, code, failure
+                )
+        except (IOError, OSError) as e:
+            code, failure = e
             raise PathError(
-                b"Path {path} is not a directory.".format(path=path.path)
+                b"Unable to write certificate file.",
+                e.filename, code, failure
             )
+        finally:
+            os.umask(original_umask)
 
-        certPath = path.child(certificate_filename)
-        keyPath = path.child(key_filename)
 
-        if not certPath.isfile():
-            raise PathError(
-                b"Certificate file {path} does not exist.".format(
-                    path=certPath.path)
-            )
+class ControlCredential(FlockerCredential):
```

Comment:
:question: Rather than subclassing, it would be better to have a `FlockerCredential` instance as an attribute. In general subclassing should be avoided if at all possible.